motionplan/inverseKinematics.go

```diff
@@ -2,23 +2,260 @@ package motionplan
+		MinScore:     defaultMinIkScore,
+		MaxSolutions: defaultSolutionsToSeed,
+		NumThreads:   defaultNumThreads,
+	}
+
+	opts.AddConstraint(defaultJointConstraint, NewJointConstraint(math.Inf(1)))
+	return opts
+}
+
+func copyIKOptions(toCopy *ikOptions) *ikOptions {
+	return &ikOptions{
+		constraintHandler: toCopy.constraintHandler,
+		extra:             deepAtomicCopyMap(toCopy.extra),
+		metric:            toCopy.metric,
+		MinScore:          toCopy.MinScore,
+		MaxSolutions:      toCopy.MaxSolutions,
+		NumThreads:        toCopy.NumThreads,
+	}
+}
+
+// inverseKinematicsSolver defines an interface which is used to solve inverse kinematics queries.
+type inverseKinematicsSolver interface {
+	solve(context.Context, chan<- []referenceframe.Input, spatialmath.Pose, []referenceframe.Input, Metric, int) error
+	frame() referenceframe.Frame
+	options() *ikOptions
+}
+
+type ikSolver struct {
+	model  referenceframe.Frame
+	logger golog.Logger
+	opts   *ikOptions
+}
+
+func (ik *ikSolver) frame() referenceframe.Frame {
+	return ik.model
+}
+
+func (ik *ikSolver) options() *ikOptions {
+	return ik.opts
+}
+
+// NewIKSolver instantiates an InverseKinematicsSolver according to the configuration parameters defined in ikConfig.
+func newIKSolver(frame referenceframe.Frame, logger golog.Logger, ikConfig map[string]interface{}) (inverseKinematicsSolver, error) {
+	// Start with normal options
+	opt := newBasicIKOptions()
+	opt.extra = ikConfig
+
+	// convert map to json, then to a struct, overwriting present defaults
+	jsonString, err := json.Marshal(ikConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(jsonString, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	// infer IK solver to build based on number of threads allowed
+	if opt.NumThreads <= 0 {
+		opt.NumThreads = defaultNumThreads
+	}
+	if opt.NumThreads == 1 {
+		return newNLOptIKSolver(frame, logger, opt)
+	}
+	return newEnsembleIKSolver(frame, logger, opt)
+}
+
+// BestIKSolutions takes an InverseKinematicsSolver and a goal location and calculates a number of solutions to achieve this goal, scored
+// by proximity to some reference input that is also specified by the user.  Finally, a WorldState argument allows users to
+// disallow or allow regions of state space through defining obstacles or interaction spaces respectively.
+func BestIKSolutions(
+	ctx context.Context,
+	frame referenceframe.Frame,
+	logger golog.Logger,
+	goal spatialmath.Pose,
+	input []referenceframe.Input,
+	worldState *referenceframe.WorldState,
+	ikConfig map[string]interface{},
+	randseed int,
+	nSolutions int,
+) ([][]referenceframe.Input, error) {
+	// Build an inverseKinematicsSolver object to operate on
+	ik, err := newIKSolver(frame, logger, ikConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build an ephemeral framesystem and make a map of the inputs to it
+	fs := referenceframe.NewEmptySimpleFrameSystem("temp")
+	if err := fs.AddFrame(frame, fs.Frame(referenceframe.World)); err != nil {
```

Comment:
We can't assume this. We need to construct the frame system using the additional transforms in WorldState, place obstacles accordingly based on that. Specifying the parentage of the frame we're solving for might be needed as well.

Reply:
For example, in the builtin motion service, the thing being solved for can be parented to things defined in additional transforms.

Reply:
The changes I've just introduced address this